### PR TITLE
feat: check configuration

### LINKED ISSUE #1813: kube-ovn-controller should exit if  default/join subnets use linklocal or muticast cidr
We should check the default/join subnet cidr at install and kube-ovn-controller init and break to prevent further issues.

### PR DESCRIPTION
**Pull request by lut777**
- [x] Make sure you have followed [Kube-OVN Code Style](https://github.com/kubeovn/kube-ovn/blob/master/CODE_STYLE.md).

#### What type of this PR
Examples of user facing changes:
- Features

Now controller will check network CIDRs before initialization, and illegal CIDR will lead to fatal error.

#### Which issue(s) this PR fixes:
Fixes #1813

### CODE REVIEW

**Comment by oilbeater** [comment]
`util/validator.go` has similar logical and should reuse it.

**Comment by lut777** [comment]
> `util/validator.go` has similar logical and should reuse it.

It's done.

File: pkg/controller/config.go [hunk 1]
```diff
@@ -249,6 +249,10 @@ func ParseFlags() (*Configuration, error) {
 		return nil, err
 	}
 
+	if err := util.CIDRConflict([]string{config.NodeSwitchCIDR, config.DefaultCIDR, config.ServiceClusterIPRange}); err != nil {
```
Comment (oilbeater) [inline pkg/controller/config.go:252]: cidrCheck 和 conflictCheck 要拆开

File: pkg/util/net.go [hunk 2]
```diff
@@ -94,23 +108,38 @@ func LastIP(subnet string) (string, error) {
...
+func CIDRConflict(subnets []string) error {
+
+	v4list := make([]*net.IPNet, 0)
+	v6list := make([]*net.IPNet, 0)
+	for _, subnet := range subnets {
+		v4, v6, err := ReturnCidr(subnet)
+		if err != nil {
+			return err
+		}
+		if v4 != nil {
```
Comment (oilbeater) [inline pkg/util/net.go:120]: 不要在 cidrconflict 里检查 cidr 是否正确。否则一个子网非法，其他正常子网都会冲突无法创建

File: pkg/util/net.go [hunk 3]
```diff
@@ -434,3 +463,100 @@ func GatewayContains(gatewayNodeStr, gateway string) bool {
...
+			last := firstLastIP[j][1]
+			if s.Contains(first) || s.Contains(last) {
+				return fmt.Errorf("%s overlaps with %s", subnets[j].String(), s.String())
+			}
+		}
+	}
+	return nil
+}
+
+func ReturnCidr(address string) (*net.IPNet, *net.IPNet, error) {
```
Comment (oilbeater) [inline pkg/util/net.go:491]:
```suggestion
func ReturnCidr(cidrBlock string) (*net.IPNet, *net.IPNet, error) {
```

File: pkg/util/net.go [hunk 4]
```diff
@@ -434,3 +463,100 @@ func GatewayContains(gatewayNodeStr, gateway string) bool {
...
+	if err != nil {
+		return nil, nil, fmt.Errorf("CIDR %v parsing failed", ips[0])
+	}
+	if ip.To4() != nil {
+		return acidr, nil, nil
+	} else if ip.To16() != nil {
+		return nil, acidr, nil
+	}
+
+	return nil, nil, fmt.Errorf("ip %v unresolvable", address)
```
Comment (oilbeater) [inline pkg/util/net.go:515]:
```suggestion
	return nil, nil, fmt.Errorf("cidr %v unresolvable", address)
```

File: pkg/util/net.go [hunk 5]
```diff
@@ -434,3 +463,100 @@ func GatewayContains(gatewayNodeStr, gateway string) bool {
...
+}
+
+func checkIPV4(ip *net.IPNet) error {
```
Comment (oilbeater) [inline pkg/util/net.go:518]:
```suggestion
func checkIPV4Cidr(cidr *net.IPNet) error {
```

File: pkg/util/net.go [hunk 6]
```diff
@@ -434,3 +463,100 @@ func GatewayContains(gatewayNodeStr, gateway string) bool {
...
+		return fmt.Errorf("%s conflict with v4 localnet cidr %s", ip.String(), IPv4zero)
+	}
+	_, v4, _ = net.ParseCIDR(IPv4LinkLocalUnicast)
+	if err := checkCIDROverlap([]*net.IPNet{ip, v4}); err != nil {
+		return fmt.Errorf("%s conflict with v4 locallink cidr %s", ip.String(), IPv4LinkLocalUnicast)
+	}
+	return nil
+}
+
+func checkIPV6(ip *net.IPNet) error {
```
Comment (oilbeater) [inline pkg/util/net.go:543]:
```suggestion
func checkIPV6Cidr(cidr *net.IPNet) error {
```

**Review by oilbeater — CHANGES_REQUESTED** [review]

File: pkg/util/net.go (hunk 6 above)
Comment (lut777) [reply pkg/util/net.go]: done

File: pkg/util/net.go (hunk 5 above)
Comment (lut777) [reply pkg/util/net.go]: done

File: pkg/util/net.go (hunk 4 above)
Comment (lut777) [reply pkg/util/net.go]: done

File: pkg/util/net.go (hunk 3 above)
Comment (lut777) [reply pkg/util/net.go]: done

File: pkg/util/net.go (hunk 2 above)
Comment (lut777) [reply pkg/util/net.go:120]: done

File: pkg/controller/config.go (hunk 1 above)
Comment (lut777) [reply pkg/controller/config.go]: 两个功能性函数已经拆开, 并且已经根据之前的函数功能重新简化命名并使用.

File: .github/workflows/build-x86-image.yaml [hunk 7]
```diff
@@ -615,6 +615,9 @@ jobs:
           command: |
             sudo make kind-install-ipv6
 
+      - name: Setup upterm session
+        uses: lhotari/action-upterm@v1
+
```
Comment (zhangzujian) [inline .github/workflows/build-x86-image.yaml:620]: should be removed

Comment (lut777) [reply .github/workflows/build-x86-image.yaml]: deleted after debug

File: pkg/util/net.go [hunk 8]
```diff
@@ -434,3 +428,77 @@ func GatewayContains(gatewayNodeStr, gateway string) bool {
 func JoinHostPort(host string, port int32) string {
 	return net.JoinHostPort(host, strconv.FormatInt(int64(port), 10))
 }
+
+func CheckSupCIDROverlap(a, b string) bool {
```
Comment (oilbeater) [inline pkg/util/net.go:432]: CheckSupCIDROverlap -> CIDROverlap

File: pkg/util/net.go [hunk 9]
```diff
@@ -434,3 +428,77 @@ func GatewayContains(gatewayNodeStr, gateway string) bool {
...
+			}
+			if aIpNet.Contains(bIp) || bIpNet.Contains(aIp) {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func CheckCIDRSpec(cidr string) error {
```
Comment (oilbeater) [inline pkg/util/net.go:451]: CheckCIDRSpec -> CIDRGlobalUnicast

File: pkg/util/net.go [hunk 10]
```diff
@@ -15,6 +15,19 @@ import (
 	kubeovnv1 "github.com/kubeovn/kube-ovn/pkg/apis/kubeovn/v1"
 )
 
+const (
+	IPV4Multicast        = "224.0.0.0/4"
+	IPV4Loopback         = "127.0.0.1/8"
+	IPv4bcast            = "255.255.255.255/32"
```
Comment (oilbeater) [inline pkg/util/net.go:21]: IPv4Broadcast

File: pkg/util/net.go [hunk 11]
```diff
@@ -15,6 +15,19 @@ import (
...
+	IPv4zero             = "0.0.0.0/32"
```
Comment (oilbeater) [inline pkg/util/net.go:22]: IPv4Zero

File: pkg/util/net.go [hunk 12]
```diff
@@ -15,6 +15,19 @@ import (
...
+	IPv4LinkLocalUnicast = "169.254.0.0/16"
+
+	IPv6unspecified      = "::/128"
+	IPv6loopback         = "::1/128"
+	IPV6Multicast        = "ff00::/8"
+	IPv6linkLocalUnicast = "FE80::/10"
```
Comment (oilbeater) [inline pkg/util/net.go:28]: All the variables should use camelcase

File: pkg/util/net.go [hunk 13]
```diff
@@ -434,3 +428,77 @@ func GatewayContains(gatewayNodeStr, gateway string) bool {
...
+			}
+		}
+	}
+	return false
+}
+
+func CheckCIDRSpec(cidr string) error {
+	for _, cidrBlock := range strings.Split(cidr, ",") {
+		if CheckSupCIDROverlap(cidrBlock, IPv4bcast) {
+			return fmt.Errorf("%s conflict with v4 boardcast cidr %s", cidr, IPv4bcast)
```
Comment (oilbeater) [inline pkg/util/net.go:454]: broadcast

File: pkg/util/net.go [hunk 14]
```diff
@@ -434,3 +428,77 @@ func GatewayContains(gatewayNodeStr, gateway string) bool {
...
+			return fmt.Errorf("%s conflict with v4 multicast cidr %s", cidr, IPV4Multicast)
+		}
+		if CheckSupCIDROverlap(cidrBlock, IPV4Loopback) {
+			return fmt.Errorf("%s conflict with v4 loopback cidr %s", cidr, IPV4Loopback)
+		}
+		if CheckSupCIDROverlap(cidrBlock, IPv4zero) {
+			return fmt.Errorf("%s conflict with v4 localnet cidr %s", cidr, IPv4zero)
+		}
+		if CheckSupCIDROverlap(cidrBlock, IPv4LinkLocalUnicast) {
+			return fmt.Errorf("%s conflict with v4 locallink cidr %s", cidr, IPv4LinkLocalUnicast)
```
Comment (oilbeater) [inline pkg/util/net.go:466]: link local

File: pkg/util/net.go [hunk 15]
```diff
@@ -434,3 +428,77 @@ func GatewayContains(gatewayNodeStr, gateway string) bool {
...
+			return fmt.Errorf("%s conflict with v6 unspecified cidr %s", cidr, IPv6unspecified)
+		}
+		if CheckSupCIDROverlap(cidrBlock, IPv6loopback) {
+			return fmt.Errorf("%s conflict with v6 loopback cidr %s", cidr, IPv6loopback)
+		}
+		if CheckSupCIDROverlap(cidrBlock, IPV6Multicast) {
+			return fmt.Errorf("%s conflict with vv6 multicast cidr %s", cidr, IPV6Multicast)
+		}
+		if CheckSupCIDROverlap(cidrBlock, IPv6linkLocalUnicast) {
+			return fmt.Errorf("%s conflict with v6 locallink cidr %s", cidr, IPv6linkLocalUnicast)
```
Comment (oilbeater) [inline pkg/util/net.go:479]: link local

File: pkg/util/net.go [hunk 16]
```diff
@@ -434,3 +428,77 @@ func GatewayContains(gatewayNodeStr, gateway string) bool {
...
+		}
+	}
+	return nil
+}
+
+func CheckCIDRsAll(cidrs []string) error {
```
Comment (oilbeater) [inline pkg/util/net.go:485]: CheckCIDRsAll -> CheckSystemCIDR

File: pkg/util/net.go [hunk 17]
```diff
@@ -434,3 +428,77 @@ func GatewayContains(gatewayNodeStr, gateway string) bool {
...
+			break
+		}
+		c := i + 1
+		for c < lenth {
+			if CheckSupCIDROverlap(cidr, cidrs[c]) {
+				err := fmt.Errorf("cidr %s is conflict with cidr %s", cidr, cidrs[c])
+				return err
+			}
+			c += 1
+		}
```
Comment (oilbeater) [inline pkg/util/net.go:501]:
```suggestion
for j := range cidrs {
    if j == i {
       continue
    }
   if CheckSupCIDROverlap(cidr, cidrs[c]) {
       err := fmt.Errorf("cidr %s is conflict with cidr %s", cidr, cidrs[c])
        return err
}
}
```

**Review by oilbeater — CHANGES_REQUESTED** [review]

File: pkg/util/net.go (hunk 8 above)
Comment (lut777) [reply pkg/util/net.go]: done

File: pkg/util/net.go (hunk 9 above)
Comment (lut777) [reply pkg/util/net.go]: done

File: pkg/util/net.go (hunk 10 above)
Comment (lut777) [reply pkg/util/net.go]: done

File: pkg/util/net.go (hunk 11 above)
Comment (lut777) [reply pkg/util/net.go]: done

File: pkg/util/net.go (hunk 12 above)
Comment (lut777) [reply pkg/util/net.go]: done

File: pkg/util/net.go (hunk 13 above)
Comment (lut777) [reply pkg/util/net.go]: done

File: pkg/util/net.go (hunk 14 above)
Comment (lut777) [reply pkg/util/net.go]: done

File: pkg/util/net.go (hunk 15 above)
Comment (lut777) [reply pkg/util/net.go]: done

File: pkg/util/net.go (hunk 16 above)
Comment (lut777) [reply pkg/util/net.go]: done

File: pkg/util/net.go (hunk 17 above)
Comment (lut777) [reply pkg/util/net.go]: fixed and done

File: pkg/util/net.go [hunk 18]
```diff
@@ -15,6 +15,19 @@ import (
 	kubeovnv1 "github.com/kubeovn/kube-ovn/pkg/apis/kubeovn/v1"
 )
 
+const (
+	IPV4Multicast        = "224.0.0.0/4"
+	IPV4Loopback         = "127.0.0.1/8"
```
Comment (oilbeater) [inline pkg/util/net.go:20]: V4 -> v4

File: pkg/util/net.go [hunk 19]
```diff
@@ -15,6 +15,19 @@ import (
...
+	IPv4Broadcast        = "255.255.255.255/32"
+	IPv4Zero             = "0.0.0.0/32"
+	IPv4LinkLocalUnicast = "169.254.0.0/16"
+
+	IPv6Unspecified      = "::/128"
+	IPv6Loopback         = "::1/128"
+	IPV6Multicast        = "ff00::/8"
```
Comment (oilbeater) [inline pkg/util/net.go:27]: V6 -> v6

File: pkg/util/net.go [hunk 20]
```diff
@@ -434,3 +428,74 @@ func GatewayContains(gatewayNodeStr, gateway string) bool {
...
+		}
+
+		if CIDROverlap(cidrBlock, IPv6Unspecified) {
+			return fmt.Errorf("%s conflict with v6 unspecified cidr %s", cidr, IPv6Unspecified)
+		}
+		if CIDROverlap(cidrBlock, IPv6Loopback) {
+			return fmt.Errorf("%s conflict with v6 loopback cidr %s", cidr, IPv6Loopback)
+		}
+		if CIDROverlap(cidrBlock, IPV6Multicast) {
+			return fmt.Errorf("%s conflict with vv6 multicast cidr %s", cidr, IPV6Multicast)
```
Comment (oilbeater) [inline pkg/util/net.go:476]: vv6 -> v6

File: pkg/controller/config.go [hunk 21]
```diff
@@ -249,6 +249,10 @@ func ParseFlags() (*Configuration, error) {
 		return nil, err
 	}
 
+	if err := util.CheckSystemCIDR([]string{config.NodeSwitchCIDR, config.DefaultCIDR, config.ServiceClusterIPRange}); err != nil {
+		return nil, err
```
Comment (oilbeater) [inline pkg/controller/config.go:253]: klog.Errorf("check system cidr failed, %v", err)

**Review by oilbeater — CHANGES_REQUESTED** [review]

File: pkg/util/net.go (hunk 18 above)
Comment (lut777) [reply pkg/util/net.go:20]: done

File: pkg/util/net.go (hunk 19 above)
Comment (lut777) [reply pkg/util/net.go:27]: done

File: pkg/util/net.go (hunk 20 above)
Comment (lut777) [reply pkg/util/net.go:476]: done

File: pkg/controller/config.go (hunk 21 above)
Comment (lut777) [reply pkg/controller/config.go:253]: added

**Review by oilbeater — APPROVED** [review]

**Pull request merged**